File: app/api/routes/analyze_transactions.py
"""Analyze-transactions route."""

from __future__ import annotations

from fastapi import APIRouter

from contracts.python.contracts import AnalyzeTransactionsRequest, AnalyzeTransactionsResponse

from app.services.emissions_service import emissions_service

router = APIRouter()


@router.post(
    "/api/analyze-transactions",
    response_model=AnalyzeTransactionsResponse,
    response_model_by_alias=True,
)
def analyze_transactions(request: AnalyzeTransactionsRequest) -> AnalyzeTransactionsResponse:
    return emissions_service.analyze_transactions(request)

File: api/src/routes/analyzeTransactions.ts
import { Router, type Request, type Response } from "express";
import {
  AnalyzeTransactionsRequestSchema,
  AnalyzeTransactionsResponseSchema,
} from "@carboniq/contracts";
import { emissionsService } from "../services/emissionsService.js";
import { getZodLikeDetails, isZodLikeError } from "../lib/validation.js";

export const analyzeTransactionsRouter = Router();

analyzeTransactionsRouter.post("/", async (req: Request, res: Response) => {
  try {
    const request = AnalyzeTransactionsRequestSchema.parse(req.body);
    const response = AnalyzeTransactionsResponseSchema.parse(
      emissionsService.analyzeTransactions(request)
    );
    res.json(response);
  } catch (err) {
    if (isZodLikeError(err)) {
      res.status(400).json({
        error: "Validation error",
        details: getZodLikeDetails(err),
      });
      return;
    }
    console.error("Analyze transactions error:", err);
    res.status(500).json({ error: "Internal server error" });
  }
});
